cmake/org.eclipse.cdt.cmake.ui/src/org/eclipse/cdt/cmake/ui/internal/CMakePreferencePage.java

```diff
@@ -141,11 +177,143 @@ public void widgetSelected(SelectionEvent e) {
+		checkBoxComp.setLayout(new GridLayout());
+
+		Button useCMakeToolLocCheckBox = new Button(checkBoxComp, SWT.CHECK);
+		useCMakeToolLocCheckBox.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, false, 3, 1));
+		useCMakeToolLocCheckBox.setText(Messages.CMakePreferencePage_UseCMakeToolLocationsInCMakeBuilds);
+		useCMakeToolLocCheckBox.setToolTipText(Messages.CMakePreferencePage_UseCMakeToolLocationsInCMakeBuildsTooltip);
+		useCMakeToolLocCheckBox.setSelection(useCmakeToolLocation);
+		useCMakeToolLocCheckBox.addListener(SWT.Selection, e -> {
+			useCmakeToolLocation = useCMakeToolLocCheckBox.getSelection();
+			updateCMakeGroup(useCmakeToolLocation);
+		});
+
+		// Set width hint to avoid Preference page opens with a horizontal scroll bar when
+		// location in textBox got too long.
+		GridData gd = new GridData(SWT.FILL, SWT.FILL, false, false);
+		gd.widthHint = control.getSize().x;
+
+		Composite locationComp = new Composite(cmakeToolsGroup, SWT.NONE);
+		locationComp.setLayoutData(gd);
+		locationComp.setLayout(new GridLayout(3, false));
+
+		Label cmakeLocationLabel = new Label(locationComp, SWT.NONE);
+		cmakeLocationLabel.setText(Messages.CMakePreferencePage_CMakeLocation);
+		cmakeLocationLabel.setToolTipText(Messages.CMakePreferencePage_CMakeLocationTooltip);
+
+		cmakeLocationTextBox = new Text(locationComp, SWT.BORDER);
+		cmakeLocationTextBox.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, true, true));
+		cmakeLocationTextBox.setText(cmakeLocation);
+		cmakeLocationTextBox.addModifyListener(new ModifyListener() {
+			@Override
+			public void modifyText(ModifyEvent evt) {
+				cmakeLocation = cmakeLocationTextBox.getText().trim();
+				testButton.setEnabled(useCmakeToolLocation && !cmakeLocation.isBlank());
+			}
+		});
+
+		Composite cmakeLocationButtonComp = new Composite(locationComp, SWT.NONE);
+		cmakeLocationButtonComp.setLayoutData(new GridData(SWT.RIGHT, SWT.CENTER, false, false));
+		cmakeLocationButtonComp.setLayout(new GridLayout(3, true));
+
+		variablesButton = new Button(cmakeLocationButtonComp, SWT.PUSH);
+		variablesButton.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, false));
+		variablesButton.setText(Messages.CMakePreferencePage_Variables);
+		variablesButton.addListener(SWT.Selection, e -> {
+			String variable = getVariableDialog(getShell(), null);
+			if (variable != null) {
+				cmakeLocationTextBox.insert(variable);
+				cmakeLocation = cmakeLocationTextBox.getText().trim();
+			}
+		});
+
+		testButton = new Button(cmakeLocationButtonComp, SWT.PUSH);
+		testButton.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, false));
+		testButton.setText(Messages.CMakePreferencePage_Test);
+		testButton.setToolTipText(Messages.CMakePreferencePage_TestTooltip);
+		testButton.addListener(SWT.Selection, e -> {
+			try {
+				Process p = Runtime.getRuntime().exec(new String[] {
+						resolveVariableValue(cmakeLocation) + File.separatorChar + "cmake", "--version" }); //$NON-NLS-1$ //$NON-NLS-2$
```

Comment:
If resolveVariableValue returns null here I think this will mean that `null/cmake` will be the result.